crates/driver/src/domain/competition/mod.rs

```diff
@@ -39,79 +40,130 @@ pub struct Competition {
     pub simulator: Simulator,
     pub now: time::Now,
     pub mempools: Vec<Mempool>,
-    pub settlement: Mutex<Option<(solution::Id, settlement::Verified)>>,
+    pub settlement: Mutex<Option<settlement::Verified>>,
 }
 
 impl Competition {
     /// Solve an auction as part of this competition.
-    pub async fn solve(&self, auction: &Auction) -> Result<(solution::Id, solution::Score), Error> {
+    pub async fn solve(
+        &self,
+        auction: &Auction,
+    ) -> Result<(settlement::Id, solution::Score), Error> {
+        // Fetch the liquidity needed by the auction.
         tracing::trace!("fetching liquidity");
-        let liquidity = self.liquidity.fetch(&Self::liquidity_pairs(auction)).await;
+        let liquidity = self
+            .liquidity
+            .fetch(
+                &auction
+                    .orders
+                    .iter()
+                    .filter_map(|order| match order.kind {
+                        order::Kind::Market | order::Kind::Limit { .. } => {
+                            liquidity::TokenPair::new(order.sell.token, order.buy.token)
+                        }
+                        order::Kind::Liquidity => None,
+                    })
+                    .collect(),
+            )
+            .await;
+
+        // Fetch the solutions from the solver.
         tracing::trace!("solving");
-        let solution = self
+        let solutions = self
             .solver
             .solve(auction, &liquidity, auction.deadline.timeout(self.now)?)
             .await?;
 
-        if solution.is_empty() {
-            // Don't waste resources on simulating an empty solution.
-            return Ok((solution.id, solution::Score::zero()));
-        }
+        // Empty solutions aren't useful, so discard them.
+        let solutions = solutions.into_iter().filter(|solution| {
+            if solution.is_empty() {
+                tracing::info!(id = ?solution.id, "discarding solution: empty");
+                false
+            } else {
+                true
+            }
+        });
+
+        // Verify the solutions, encoding them into settlements.
+        let settlements = join_all(solutions.map(|solution| async move {
```

Comment:
This unbounded join_all should probably be bounded. Or at least we need to think about what the expected number of returned solutions is. If we get 1000 solutions we don't want to simulate them all concurrently because the node can't handle that.

Reply:
I would test this first.

Reply:
Didn't @nlordell take care of this problem by always using `transport::Buffered` under the hood? ([PR](https://github.com/cowprotocol/services/pull/696))

Reply:
With the right buffered transport configuration you're right that it will limit the concurrency. Still feels a bit fragile because I had to look that up and it would change if the transport config changes. I'd feel better with a limit here but it isn't that important.

Feel free to create an issue for it and not do it in this PR.

Another thought is that for multiple solutions a streaming approach is more fitting. Then Driver can start simulating immediately without waiting for all solutions and can cut off when the deadline is near. That's out of scope for this PR too but might be nice to have for the future.

Reply:
https://github.com/cowprotocol/services/issues/1473